src/inc/stackstring.h

```diff
@@ -0,0 +1,196 @@
+// Licensed to the .NET Foundation under one or more agreements. 
+// The .NET Foundation licenses this file to you under the MIT license. 
+// See the LICENSE file in the project root for more information. 
+
+
+#ifndef __STACKSTRING_H_
+#define __STACKSTRING_H_
+
+#include "malloc.h"
+
+template <SIZE_T STACKCOUNT, class T>
+class StackString
+{
+private:
+    T m_innerBuffer[STACKCOUNT + 1];
+    T * m_buffer;
+    SIZE_T m_size; // actual allocated size
+    SIZE_T m_count; // actual length of string
+
+    void NullTerminate()
+    {
+        m_buffer[m_count] = 0;
+    }
+
+    void DeleteBuffer()
+    {
+        if (m_innerBuffer != m_buffer)
+            free(m_buffer);
+
+        m_buffer = NULL;
+        return;
+    }
+
+    void ReallocateBuffer(SIZE_T count)
+    {
+        // count is always > STACKCOUNT here.
+        T * newBuffer = (T *)malloc((count + 1) * sizeof(T));
```

Comment:
For the VM, we should consider using "new" here to allow the allocation via VM's new implementation that gives hosts like SQL server a chance to override the allocation semantics and allow/disallow the allocation.